cmd/contour/serve.go

```diff
@@ -296,79 +302,106 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 		return err
 	}
 
-	// XXX(jpeach) we know the config file validated, so all
-	// the timeouts will parse. Shall we add a `timeout.MustParse()`
-	// and use it here?
+	var connectionIdleTimeout timeout.Setting
+	var streamIdleTimeout timeout.Setting
+	var delayedCloseTimeout timeout.Setting
+	var maxConnectionDuration timeout.Setting
+	var connectionShutdownGracePeriod timeout.Setting
+	var requestTimeout timeout.Setting
 
-	connectionIdleTimeout, err := timeout.Parse(ctx.Config.Timeouts.ConnectionIdleTimeout)
-	if err != nil {
-		return fmt.Errorf("error parsing connection idle timeout: %w", err)
-	}
-	streamIdleTimeout, err := timeout.Parse(ctx.Config.Timeouts.StreamIdleTimeout)
-	if err != nil {
-		return fmt.Errorf("error parsing stream idle timeout: %w", err)
-	}
-	delayedCloseTimeout, err := timeout.Parse(ctx.Config.Timeouts.DelayedCloseTimeout)
-	if err != nil {
-		return fmt.Errorf("error parsing delayed close timeout: %w", err)
-	}
-	maxConnectionDuration, err := timeout.Parse(ctx.Config.Timeouts.MaxConnectionDuration)
-	if err != nil {
-		return fmt.Errorf("error parsing max connection duration: %w", err)
+	if contourConfiguration.Envoy.Timeouts != nil {
+		if contourConfiguration.Envoy.Timeouts.ConnectionIdleTimeout != nil {
+			connectionIdleTimeout, err = timeout.Parse(*contourConfiguration.Envoy.Timeouts.ConnectionIdleTimeout)
+			if err != nil {
+				return fmt.Errorf("error parsing connection idle timeout: %w", err)
+			}
+		}
+		if contourConfiguration.Envoy.Timeouts.StreamIdleTimeout != nil {
+			streamIdleTimeout, err = timeout.Parse(*contourConfiguration.Envoy.Timeouts.StreamIdleTimeout)
+			if err != nil {
+				return fmt.Errorf("error parsing stream idle timeout: %w", err)
+			}
+		}
+		if contourConfiguration.Envoy.Timeouts.DelayedCloseTimeout != nil {
+			delayedCloseTimeout, err = timeout.Parse(*contourConfiguration.Envoy.Timeouts.DelayedCloseTimeout)
+			if err != nil {
+				return fmt.Errorf("error parsing delayed close timeout: %w", err)
+			}
+		}
+		if contourConfiguration.Envoy.Timeouts.MaxConnectionDuration != nil {
+			maxConnectionDuration, err = timeout.Parse(*contourConfiguration.Envoy.Timeouts.MaxConnectionDuration)
+			if err != nil {
+				return fmt.Errorf("error parsing max connection duration: %w", err)
+			}
+		}
+		if contourConfiguration.Envoy.Timeouts.ConnectionShutdownGracePeriod != nil {
+			connectionShutdownGracePeriod, err = timeout.Parse(*contourConfiguration.Envoy.Timeouts.ConnectionShutdownGracePeriod)
+			if err != nil {
+				return fmt.Errorf("error parsing connection shutdown grace period: %w", err)
+			}
+		}
+		if contourConfiguration.Envoy.Timeouts.RequestTimeout != nil {
+			requestTimeout, err = timeout.Parse(*contourConfiguration.Envoy.Timeouts.RequestTimeout)
+			if err != nil {
+				return fmt.Errorf("error parsing request timeout: %w", err)
+			}
+		}
```

Comment:
This is more general comment than just for this block alone: Would it make sense to try move code from `doServe()` since it starts to get really long?  
For example, if there was listener builder that takes `ContourConfiguration`, maybe the above code could be moved there from `doServe()`. Similarly maybe dag builder could take `ContourConfiguration` to offload parsing some parameters from `doServe()` to the builder.

Reply:
Yes for sure! I was trying to convert and keep the existing logic in place for easier reviewing. I can take some new commits from here and clean some of this up if it doesn't get too difficult to review.